### ISSUE TITLE
[Feature] Optimize incremental 'insert_overwrite' strategy

### ISSUE THREAD

**Issue by AxelThevenot** [dbt-labs/dbt-bigquery#1409]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-bigquery functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

The `MERGE` statement is sub-optimized in BigQuery. 
This is of course ok for unique keys as this is what we are looking for.

But for the `insert_overwrite` strategy where we are looking to rows at the partition-level there is a better solution and here is why:
- a `DELETE` or `INSERT` statement is cheapest than a `MERGE` statement.
- incremental tables are the most expensive tables in real-world projects.
- The `DELETE` statement in BigQuery is free at the partition-level.

> This has been tested at Carrefour which is my company. 
- On this replacement of the `MERGE` statement **it reduces the cost by 50.4% and the elapsed time by 35.2%** 
- On the overall procedure it reduces the cost by 26.1% and the elapsed time by 23.1%

> This is wrapped in a transaction to avoid deleting rows if any error occurs.

### Describe alternatives you've considered

I have considered an alternative to create an additional `delete+insert` incremental strategy.
But overriding the existing `insert_overwrite` is the more convenient as this is exactly what we are looking for and the features are the same.

### Who will this benefit?

Of course, **everyone without any changes.**

### Are you interested in contributing this feature?

Yes

**Comment by philBatardiere** [comment]
Yes agreed if your incremental refresh a lot of partitions back then it can takes time (usually i refresh around 3 partitions back so the process isn’t so long).

For debuging purposes, I have seen some benefits having the tmp table by restoring it with the time travel feature. But its clearly worst for auditing purposes, it should be nice to have labels related to the dbt run id in the copy config btw 😀.

I keep your feature in mind in case of future need.

Thanks!

**Comment by AxelThevenot** [comment]
> For debuging purposes, I have seen some benefits having the tmp table by restoring it with the time travel feature. But its clearly worst for auditing purposes, it should be nice to have labels related to the dbt run id in the copy config btw 😀.

Oh yes it could be nice to have those labels

So you won't valide my pull request for now from what I understand ? 

In any case, feel free to reach me if you want some help or anything else
https://www.linkedin.com/in/axel-thevenot/

**Comment by philBatardiere** [comment]
Unfortunately I don't have yet the authority to do it :) But I hope someone will do it

**Comment by AxelThevenot** [comment]
I can create an additional `delete+insert` strategy instead of overriding the `insert_overwrite`, even if it make less sense to me 👍

**Comment by amychen1776** [comment]
@AxelThevenot, I don't think I quite understand your justification for not creating a new `delete+insert` materialization strategy.  Changing a commonly used incremental strategy can cause some unintended results, especially since it seems like your intent is to remove the merge statement from the `insert_override` statement (or am I misunderstanding?)

**Comment by borjavb** [comment]
This is great stuff! But I think we should create a `delete+insert` operation instead of changing the already widely used `insert_overwrite`, as it could have unexpected consequences across all users. (+1 on @amychen1776)


For example, the `DELETE` operation being free is only under certain conditions as stated in the [documentation](https://cloud.google.com/bigquery/docs/using-dml-with-partitioned-tables#using_dml_delete_to_delete_partitions) and you might not always get the slot-free operation unfortunately. I’ve seen cases where the costs of the DELETE + INSERT end up being higher in aggregate to the whole MERGE. It’s rare, but It can happen and rolling this change for such a fundamental macro can backfire. At the end of the day is all about testing strategies that better suit the use case.

For example granular partitioned tables do not benefit from free DELETEs :why: 🤦  :
<img width="955" alt="image" src="https://github.com/user-attachments/assets/91d83e68-ef88-4b2a-9f2a-37141c529bd4">

Using a `delete+insert` will definitely be a performance boost in most cases, but users have to be aware of how to get the best of it. Better to be clear through documentation and through new methods rather than changing the current one.

**Comment by AxelThevenot** [comment]
@amychen1776 and @borjavb I got your points, I will go for a `delete+insert` strategy if this is more convenient to you and avoid having edge cases for specific users

I will make the change as soon as possible and tell you when it's done :)

**Comment by borjavb** [comment]
Been thinking about this and... how do we feel about a operator/function configuration within `instert_overwrite`?

The way I've hijacked the `insert_overwrite` strategy in the past to implement this was through the config block and it always worked really well. [Check this draft PR with the gist behind it](https://github.com/dbt-labs/dbt-bigquery/pull/1415/files).

Technically we have an incremental strategy that can work with different functions, but apply the same semantics of replacing a whole partition. So basically we can define a `incremental_strategy` plus the operator/function we want to use: either a merge or a `delete+insert`. So we get the best of both worlds:

1. If it's not defined, we default to `merge`, so we keep backwards compatibility
2. If the user selects the new `delete+insert` logic, we swap the macro that works under the hood

We already have custom configurations for merge like `merge_update_columns` and `merge_exclude_columns`, so this doesn't break with how we do things with other incremental strategies.

So the config block would look like something like this:

```
{{
    config(
        materialized="incremental",
        incremental_strategy="insert_overwrite",
        insert_overwrite_function='delete+insert',
        cluster_by="id",
        partition_by={
            "field": "date_time",
            "data_type": "datetime",
            "granularity": "day"
        },
        partitions=partitions_to_replace,
    )
}}
```

Although I'm torn if we should add the `incremental_strategy` as a dict something like this... but given that we haven't done that for the other configurations for merge... it feels now weird to add it as a dictionary? Not sure.
```
incremental_strategy={
   strategy = "insert_overwrite",
   operation = "delete+insert"
}
```
But this would just be some syntax sugar of how we want to expose the config. ButI think with this we can let the user decide which operator is the most performant, and keep full backwards compatibility.

**Comment by AxelThevenot** [comment]
@borjavb this is a great idea ! I will go for it 😄

**Comment by jtcohen6** [comment]
@AxelThevenot @borjavb Very cool to see the level of interest here :)

> But `delete+insert` is at the unique-key-level in other adapters and I want a partition-level
> Any idea of if I reuse this name even if this is not doing the same or if I give a new name ? (and what name?)

This is a totally fair point!

FWIW... The `unique_key` specified for other adapters does not actually _have_ to be unique, and many users have a configuration like this to achieve the outcome you're after here:
```sql
{{ config(
   incremental_strategy = "insert_overwrite",
   unique_key = "my_partition_by_column_name",  -- not actually unique!
   ...
) }}
```

I know that nomenclature is misleading (Grace [called it out](https://github.com/dbt-labs/dbt-core/discussions/10672) a few months ago), but it would be consistent with how users have gotten this working on other DWHs/adapters. (I say that, at the same time acknowledging that non-unique keys can lead to poor query performance; there's a suggested performance optimization to switch from `delete ... using` to `delete ... where`: https://github.com/dbt-labs/dbt-adapters/issues/150#issuecomment-2048092121, https://github.com/dbt-labs/dbt-adapters/pull/151.)

---

To offer one other perspective:
- We're rolling out a **brand-new `microbatch` incremental strategy** in dbt Core v1.9 (https://docs.getdbt.com/docs/build/incremental-microbatch)
- This strategy has parallelism _built in_, with the capability within dbt for per-batch concurrency + retry
- Our stated goal: "To perform this operation, dbt will use the most efficient atomic mechanism for "full batch" replacement that is available on each data platform." Right now, that's using `merge` with a constant-false predicate — the same as `insert_overwrite` ([here](https://github.com/dbt-labs/dbt-bigquery/blob/83bb413d770fa1adb68cdb738dde0351347811ad/dbt/include/bigquery/macros/materializations/incremental_strategy/microbatch.sql#L20) — but with the option to use the `copy_partitions` API instead ([as Borja noted](https://github.com/dbt-labs/dbt-bigquery/issues/1414#issuecomment-2505751621)).
- Given that `microbatch` is brand-new — should we make it use a `delete; insert` transaction right from the start? Or, should we add this as another opt-in config for the `insert_overwrite` strategy — one that would also be implicitly available to the `microbatch` strategy, too)? It's probably too late in the game to be asking, because dbt Core v1.9 is planning to go live in early December, but I figured I'd ask anyway :)
